rfcs/027-pipeline-intermediate-storage/README.md

```diff
@@ -0,0 +1,147 @@
+
+### Redis
+
+Pros:
+
+* In memory so likely to be fastest reads / writes out of all the options.
+
+* AWS offers it via ElastiCache.
+
+* Improved visibility over VHS.
+
+Cons:
+
+* Requires an instance with enough memory to store all the data, so whilst certainly possible with our current sizes of data the point at which it becomes infeasible will be reached much quicker than other solutions.
+
+* No inbuilt versioning so would have to implement something manually within our applications.
+
+* Visibility nothing like Kibana, and not particularly queryable.
+
+Questions:
+
+* Redis offers some persistence, either via regular snapshots or by logging individual create / update / delete operations to disk with the ability to reconstruct the store via playback. Would need further investigation.
+
+* Pricing is unknown, likely would need some pretty large reserved EC2 instances but expect it to be cheaper than DynamoDB / S3 combination.
+
+### AWS RDS / PostgreSQL
+
+Pros:
+
+* The most reliable and predictable option due to being the only solution which is completely ACID compliant.
+
+* Can support our versioning needs (this could be achieved e.g. within our application logic using transactions, or within a more complex single query using clauses).
+
+* Available as a service on AWS.
+
+* Improved visibility over VHS, and ability to query data.
+
+* Replicas can help with scalability when large number of reads.
+
+* Transaction support might be useful for some of our current logic or future use cases.
+
+Cons:
+
+* Performance may not match other options due to stronger consistency guarantees (though likely to be much better than VHS due to only hitting a single system and making use of a cache).
+
+* Writing raw SQL is not fun, and I have not really been satisfied with any of the Scala wrappers I have tried.
+
+* Visibility nothing like Kibana (there are some web interfaces available which attempt to do similar things, but would require more manual setup and maybe not as polished).
+
+Questions:
+
+* Costs?
+
+* Will not scale as well to mega large data sets: probably not an issue with the sizes of data we are likely to deal with, but worth considering.
+
+## Implementation
+
+Orthogonal to the choice of store, the way we write data into the store for usage later in the pipeline has two main possibilities.
+
+### Service
+
+The current `recorder` service is a wrapper round VHS which stores items received at its input, emitting a message contained the ID of the stored item for downstream pipeline stages to use themselves for access in future.
+
+ We could mirror this design using a different store to VHS for upcoming intermediate storage needs. As the behaviour would be essentially the same regardless of what state the works were in (unmerged, merged, enriched etc) the exact same service can be inserted in different parts of the pipeline differing only in configuration of what store the service has access to.
```

Comment:
Another example of this would be the `ingestor`s?

Reply:
Yes but I think the ingestors do more than what we require for intermediate storage (creating mappings etc), so might not be helpful to mix this up together.

Reply:
mmmh I think the `ingestor_common` library wants you to define an `IndexConfig `for your ingestor, but you can create an `IndexConfig` that is jus dynamic mapping. It might be simpler to use the ingestor for new storage needs rather than the recorder..